drivers/volume/portworx/portworx.go

```diff
@@ -2438,6 +2438,75 @@ func (p *portworx) DeletePair(pair *storkapi.ClusterPair) error {
 	return nil
 }
 
+func (p *portworx) Failover(action *storkapi.Action) error {
+	namespace := action.Namespace
+	logrus.Debugf("Failover: namespace %s", namespace)
+	volDriver, err := p.getAdminVolDriver() // TODO(dgoel): admin or user vol driver?
+	if err != nil {
+		return err
+	}
+
+	// TODO(dgoel): when should we return error
+	// and should we just log an error and continue
+
+	// TODO(dgoel): show relevant events in the Action CR
+	// return list of errors (as above) - controller can use to log events
+	// for ex: a volume promote failed
+
+	// can we directly fetch PersistentVolumes here?
+	pvcList, err := core.Instance().GetPersistentVolumeClaims(namespace, nil)
+	if err != nil {
+		return fmt.Errorf("error getting list of volumes to migrate: %v", err)
+	}
+	for _, pvc := range pvcList.Items {
+		if !p.OwnsPVC(core.Instance(), &pvc) {
+			continue
+		}
+		if resourcecollector.SkipResource(pvc.Annotations) { // TODO(dgoel): do we need this?
+			continue
+		}
+
+		pvName, err := core.Instance().GetVolumeForPersistentVolumeClaim(&pvc)
+		if err != nil {
+			return fmt.Errorf("error getting volume for PVC: %v", err)
+		}
+		logrus.Debugf("Failover: PV %v", pvName)
+
+		pv, err := core.Instance().GetPersistentVolume(pvName)
+		if err != nil {
+			return err
+		}
+		volID := pv.Spec.PortworxVolume.VolumeID
+		if volID == "" {
+			return fmt.Errorf("PV contains \"\" value for Spec.PortworxVolume.VolumeID")
+		}
+		logrus.Debugf("Failover: volID %v", volID)
+
+		volumes, err := volDriver.Inspect([]string{volID})
```

Comment:
You can query Portworx with pvName as well
```suggestion
		volumes, err := volDriver.Inspect([]string{pvName})
```

Reply:
Is it safe to assume that Px volume name will always be the same as PV name?

Reply:
Yes the PV name needs to be either the PX volume ID or volumeName. The Inspect() API accepts both PX volumeID and volumeName.